consensus/types/src/beacon_state/tree_hash_cache.rs

```diff
@@ -21,9 +22,66 @@ const NODES_PER_VALIDATOR: usize = 15;
 /// Do not set to 0.
 const VALIDATORS_PER_ARENA: usize = 4_096;
 
+#[derive(Debug, PartialEq, Clone, Encode, Decode)]
+pub struct Eth1DataVotesTreeHashCache<T: EthSpec> {
+    arena: CacheArena,
+    tree_hash_cache: TreeHashCache,
+    voting_period: u64,
+    roots: VariableList<Hash256, T::SlotsPerEth1VotingPeriod>,
+}
+
+impl<T: EthSpec> Eth1DataVotesTreeHashCache<T> {
+    /// Instantiates a new cache.
+    ///
+    /// Allocates the necessary memory to store all of the cached Merkle trees but does perform any
+    /// hashing.
+    pub fn new(state: &BeaconState<T>) -> Self {
+        let mut arena = CacheArena::default();
+        let roots: VariableList<_, _> = state
+            .eth1_data_votes
+            .iter()
+            .map(|eth1_data| eth1_data.tree_hash_root())
+            .collect::<Vec<_>>()
+            .into();
+        let tree_hash_cache = roots.new_tree_hash_cache(&mut arena);
+
+        Self {
+            arena,
+            tree_hash_cache,
+            voting_period: Self::voting_period(state.slot),
+            roots,
+        }
+    }
+
+    fn voting_period(slot: Slot) -> u64 {
+        slot.as_u64() / T::SlotsPerEth1VotingPeriod::to_u64()
+    }
+
+    pub fn recalculate_tree_hash_root(&mut self, state: &BeaconState<T>) -> Result<Hash256, Error> {
+        if state.eth1_data_votes.len() < self.roots.len()
+            || Self::voting_period(state.slot) != self.voting_period
+        {
+            std::mem::replace(self, Self::new(state));
```

Comment:
I think this will be disallowed by CI when you merge master, because it can just be `*self = Self::new(state)`